kronofoto/static/assets/js/lib.js

```diff
@@ -18,633 +18,808 @@ import { Tooltip } from 'foundation-sites/js/foundation.tooltip';
-  let currentPosition = preItemNum + quantizedPositionX + 1
 
-  let numThumbnails= $('#fi-thumbnail-carousel-images li').length
+  initDraggableThumbnails() {
+    let self = this
+    window.$ = self.jquery
+    if (!self.jquery('#fi-thumbnail-carousel-images [data-origin]').length) {
+      self.jquery('#fi-thumbnail-carousel-images [data-active]').attr('data-origin', '')
+    }
+    self.jquery('#fi-thumbnail-carousel-images').draggable({
+      axis: 'x',
+      start: (event, ui) => {
+        self.lastActiveThumbnail = self.jquery('#fi-thumbnail-carousel-images [data-active]')
+        self.dragOffset = self.jquery('#fi-thumbnail-carousel-images').position().left
+      },
+      drag: (event, ui) => {
+        let offset = parseInt(self.jquery('#fi-thumbnail-carousel-images').attr('data-offset')) || 0
+        if(offset) {
+          self.jquery('#fi-thumbnail-carousel-images').attr('data-offset', 0)
+        }
+        ui.position.left = ui.position.left - offset
+        self.moveTimelineCoin(ui.position.left)
+      },
+      stop: (event, ui) => {
+        self.dragOffset = false
+        self.lastActiveThumbnail = null
+        let offset = parseInt(self.jquery('#fi-thumbnail-carousel-images').attr('data-offset')) || 0
+        if(offset > 0) {
+          self.jquery('#fi-thumbnail-carousel-images').attr('data-offset', 0)
+        }
+        ui.position.left = ui.position.left - offset
+        // self.htmx.trigger(self.jquery('#fi-thumbnail-carousel-images li[data-active] span').get(0), 'manual')
+        self.dropTimelineCoin(ui.position.left)
+      }
+    })
+    setTimeout(() => {
+      self.jquery('#fi-thumbnail-carousel-images').removeClass('dragging')
+      if (!self.jquery('#fi-thumbnail-carousel-images [data-origin]').length) {
+        self.jquery('#fi-thumbnail-carousel-images [data-active]').attr('data-origin', '')
+      }
+    }, 100)
+  }
+
+  moveTimelineCoinAbsolute(pos, drag) {
+    let self = this
 
-  if(drag && numThumbnails - currentPosition < preItemNum) {
-    getMoreThumbnailsRight()
+    self.moveTimelineCoin(deltaX, drag)
   }
-  else if(drag && currentPosition < preItemNum) {
-    getMoreThumbnailsLeft()
+
+  initPopups() {
+    let self = this
+    let popups = ['add-to-list-popup', 'download-popup', 'share-popup']
+    popups.forEach((e,i) => {
+      if(self.jquery('#' + e).find('.photo-menu-popup__wrapper').length && self.jquery('#' + e).find('.photo-menu-popup__wrapper > div').length == 0) {
+        let photoId = self.jquery('#' + e + '[data-photo-id]').data('photo-id')
+        let loadedPhotoId = self.jquery('#' + e + ' [data-photo-id]').data('photo-id')
+        if(photoId != loadedPhotoId) {
+          let $placeholder = self.jquery('<div/>').attr('data-photo-id', photoId)
+          self.jquery('#' + e + ' .photo-menu-popup__wrapper').append($placeholder)
+          self.htmx.trigger(self.jquery('#' + e + ' .photo-menu-popup__wrapper').get(0), 'manual')
```

Comment:
**RE photoId/loadedPhotoId:** You're right, there is a typo here. loadedPhotoId was meant to be pulled from the hx-get variable. This approach assumes that photo_partial HTMX swap is just swapping the __wrapper popup element and it's hx-get variable. Then onLoad, initPopups runs. This function is meant to be idempotent and will only trigger a swap if the loaded id from [data-photo-id] in the swapped content doesn't match the photo-id from hx-get. **Why do this?** Something is wrong with the HTMX-swapped popups. I keep running into the issue and it's complicated and forcing a certain solution. I don't remember the details but I will note this next time I see it for us to discuss. This is also related to the git tug of war we're having over the popup blocks. I 100% agree about how it seems unnecessary. But it has a purpose. And if we can put the popup blocks back and everything works, I'd happily get rid of this function. But again, I was trying to get around an issue that emerged from that.

**RE Manual HTMX:** We've chatted about this before. This is me solutioning around HTMX. We have reason to NOT use HTMX events like intersect and click, etc. If you don't specify an hx-trigger (I think) HTMX just uses some default. We have custom events defined elsewhere like kronofoto:timelineYearChange. The point is, we define the event to be triggered from JS only. The intersect event could possibly be better used here although it causes a slight flash of content when you open the popup. I don't think there are any instances where the manual event is not handled. This should be handled by the matching .photo-menu-popup__wrapper element with hx-get and hx-trigger="manual". Am I wrong about that?

Putting the solution aside, I'm just trying to trigger an HTMX swap for a specific element "manually" and trying to ensure HTMX doesn't swap it form me with some other event like intersect or click.

Reply:
I think the issue you were trying to fix with photoId/loadedPhotoId was caused by having instantiated two instances of HTMX. I have fixed this. The click handlers are still getting installed twice somehow when switching from grid to timeline.

If the flicker is intolerable, the hx-trigger can be "load" or the contents can be put into the swap by including the template in photo_detail.html. I think it's okay to load on demand because the user is not likely to browse image to image with the share popup open. We don't need to use manual triggering here.

If the javascript is handling "click" and parsing things from hx-get attributes, making requests, and/or modifying contents of responses, these are all signs that the document is wrong. We don't need to work around our own server.